### ISSUE TITLE
 o.s.c.s.a.z.ZipkinAutoConfiguration : Check result of the [WebClientSender{https://my-zipkin:443/api/v2/spans}] contains an error [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: WebClientSender{https://my-zipkin:443/api/v2/spans} check() timed out after 1000ms}] 

### ISSUE THREAD

**Issue by patpatpat123** [spring-cloud/spring-cloud-sleuth#2126]
Hello Spring Cloud Sleuth Team,

I am reaching out because I am observing a string issue, hence would like to report it.

Setup: The application is a Spring Boot 2.6.4 + Jubilee 2021.0.1
We have another Zipkin server running fine, in my example, I call it https://my-zipkin:443
Here is my pom for the sleuth/zipkin part:

```
   <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.6.4</version>
        <relativePath/>
    </parent>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-dependencies</artifactId>
                <version>2021.0.1</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-webflux</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-cassandra-reactive</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-security</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-actuator</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-starter-kubernetes-fabric8</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-config-client</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-starter-sleuth</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-sleuth-zipkin</artifactId>
        </dependency>
        <dependency>
            <groupId>io.projectreactor.netty</groupId>
            <artifactId>reactor-netty-http-brave</artifactId>
        </dependency>
        <dependency>
            <groupId>io.micrometer</groupId>
            <artifactId>micrometer-registry-prometheus</artifactId>
        </dependency>
        <dependency>
            <groupId>io.github.resilience4j</groupId>
            <artifactId>resilience4j-micrometer</artifactId>
        </dependency>
        <dependency>
            <groupId>io.github.resilience4j</groupId>
            <artifactId>resilience4j-reactor</artifactId>
        </dependency>
        <dependency>
            <groupId>io.github.resilience4j</groupId>
            <artifactId>resilience4j-spring-boot2</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-starter-contract-verifier</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.mockito</groupId>
            <artifactId>mockito-core</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>com.lmax</groupId>
            <artifactId>disruptor</artifactId>
            <version>3.4.4</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.statemachine</groupId>
            <artifactId>spring-statemachine-core</artifactId>
            <version>3.0.1</version>
        </dependency>
        <dependency>
            <groupId>net.logstash.logback</groupId>
            <artifactId>logstash-logback-encoder</artifactId>
            <version>7.0.1</version>
        </dependency>
        <dependency>
            <groupId>de.codecentric</groupId>
            <artifactId>spring-boot-admin-starter-client</artifactId>
            <version>2.6.2</version>
        </dependency>
        <dependency>
            <groupId>org.springdoc</groupId>
            <artifactId>springdoc-openapi-webflux-ui</artifactId>
            <version>1.6.6</version>
        </dependency>
    </dependencies>
```

Reproducible 100%, on each application startup, I am seeing:

```
2022-03-05 12:16:35.477 WARN [myservice,,] 10 --- [ main] o.s.c.s.a.z.ZipkinAutoConfiguration : Check result of the [WebClientSender{https://my-zipkin:443/api/v2/spans}] contains an error [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: WebClientSender{https://zmy-zipkin:443/api/v2/spans} check() timed out after 1000ms}] 
```

The thing is, I am monitoring the Zipkin server in parallel, and it is up and running fine! Other non Spring Boot application connects to it fine during the same time stamp, etc.

What is also strange is later with the same app (the one having the warning) I do receive requests, and I am able to search the traces in the Zipkin server, proof it successfully connected to Zipkin Server, as I am able to see the full traces.

Therefore, I would like to report this issue, not understanding what is going in there. Why is there a time out where apparently, server is working fine, and client is able to send traces to it?

Is there a way to increase this time out from 1000ms to something a bit higher?

Thank you

**Comment by marcingrzejszczak** [comment]
What's your full pom setup, do you have webclient on the classpath?

**Comment by patpatpat123** [comment]
Hello @marcingrzejszczak ,

Thank you for looking into this.
I believe I go have webclient on the class path as it is a Spring WebFlux Application.
Furthermore, I a m using Webclient as http client to make requests to other parties as part of business logic.

May I ask if you are saying that having WebClient is the root cause of this issue?
Question edited with the relevant part of the pom.

Again, thank you

**Comment by marcingrzejszczak** [comment]
Duplicate of https://github.com/spring-cloud/spring-cloud-sleuth/issues/2127

**Comment by M3yo** [comment]
Hi,

We also encounters the same issue in spring boot 2.6.4 / spring cloud 2021.0.1 in a reactive app.

It seems that the bug is coming somewhere in between ZipkinAutoConfiguration.checkResult method and WebClientSender implementation. I'm not a thread expert but something is off.

The 1000ms (which is plenty, the endpoint respond in 4ms) timeout seems to be triggered even before the request is fired.

`2022-03-28 16:25:09.208  WARN [,,] 93858 --- [           main] o.s.c.s.a.z.ZipkinAutoConfiguration      : Check result of the [WebClientSender{http://localhost:9411/api/v2/spans}] contains an error [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: WebClientSender{http://localhost:9411/api/v2/spans} check() timed out after 1000ms}]`

Then a 400 error is logged:
```
2022-03-28 16:25:09.592 ERROR [,,] 93858 --- [or-http-epoll-2] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: org.springframework.web.reactive.function.client.WebClientResponseException$BadRequest: 400 Bad Request from POST http://localhost:9411/api/v2/spans
Caused by: org.springframework.web.reactive.function.client.WebClientResponseException$BadRequest: 400 Bad Request from POST http://localhost:9411/api/v2/spans
	at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:196) ~[spring-webflux-5.3.16.jar:5.3.16]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ 400 from POST http://localhost:9411/api/v2/spans [DefaultWebClient]
Original Stack Trace:
		at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:196) ~[spring-webflux-5.3.16.jar:5.3.16]
		at org.springframework.web.reactive.function.client.DefaultClientResponse.lambda$createException$1(DefaultClientResponse.java:207) ~[spring-webflux-5.3.16.jar:5.3.16]
		at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106) ~[reactor-core-3.4.15.jar:3.4.15]
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79) ~[reactor-core-3.4.15.jar:3.4.15]
...
```

@marcingrzejszczak it doesn't seems to be related to #2127 though. Can you confirm this is the same bug ?
Is there any workaround for reactive apps ? go back to sleuth 3.1.0 seems to solve the issue, but not ideal...


To reproduce:
- Start this demo [quick example](https://github.com/M3yo/sleuth-3.1.1-reactive-bug)
- and start a Zipkin server or Jaeger:

```
docker run -d --name jaeger \  
  -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 \
  -p 5775:5775/udp \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 14250:14250 \
  -p 14268:14268 \
  -p 14269:14269 \
  -p 9411:9411 \

```

**Comment by marcingrzejszczak** [comment]
Can you check the latest snapshots? I think I've fixed this already

**Comment by M3yo** [comment]
I've checked with spring-cloud-sleuth 3.1.2-SNAPSHOT (by setting springCloudVersion to 2021.0.2-SNAPSHOT) and **results are the same**.

```
2022-03-28 17:34:05.324  WARN [,,] 108992 --- [           main] o.s.c.s.a.z.ZipkinAutoConfiguration      : Check result of the [WebClientSender{http://localhost:9411/api/v2/spans}] contains an error [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: Timed out after 1000ms}]
2022-03-28 17:34:05.655  INFO [,,] 108992 --- [           main] o.s.b.web.embedded.netty.NettyWebServer  : Netty started on port 8080
2022-03-28 17:34:05.685  INFO [,,] 108992 --- [           main] com.example.demo.DemoApplication         : Started DemoApplication in 3.101 seconds (JVM running for 3.735)
2022-03-28 17:34:05.724 ERROR [,,] 108992 --- [or-http-epoll-2] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: org.springframework.web.reactive.function.client.WebClientResponseException$BadRequest: 400 Bad Request from POST http://localhost:9411/api/v2/spans
Caused by: org.springframework.web.reactive.function.client.WebClientResponseException$BadRequest: 400 Bad Request from POST http://localhost:9411/api/v2/spans
	at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:196) ~[spring-webflux-5.3.16.jar:5.3.16]
```
So the "async" issue isn't fixed, and the 400 error isn't fixed either.

For the 400 error, it seems the "Content-Type" header is the culprit.
```
curl -X POST http://localhost:9411/api/v2/spans -H "Content-Type: application/json" -d '[]' -i
HTTP/1.1 202 Accepted
Vary: Origin
Date: Mon, 28 Mar 2022 15:43:10 GMT
Content-Length: 0
```
Without header I get a 400, which sound familiar 
```
curl -X POST http://localhost:9411/api/v2/spans  -d '[]' -i 
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Vary: Origin
X-Content-Type-Options: nosniff
Date: Mon, 28 Mar 2022 15:51:39 GMT
Content-Length: 25
```

In RestTemplateSender is [contentType is correctly set](https://github.com/spring-cloud/spring-cloud-sleuth/blob/3.1.x/spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin2/RestTemplateSender.java#L49).

But it's not explicitly set in [WebclientSender](https://github.com/spring-cloud/spring-cloud-sleuth/blob/3.1.x/spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin2/WebClientSender.java#L40), only "accept" header is set.

**Comment by marcingrzejszczak** [comment]
Hmm that's interesting

**Comment by marcingrzejszczak** [comment]
I've started a Zipkin server using the latest available Zipkin server version (it's running on port `49214`). I executed the following curl request

```
curl -X POST http://localhost:49214/api/v2/spans -d '[]' -i 
HTTP/1.1 202 Accepted
content-length: 0
server: Armeria/1.13.4
date: Tue, 29 Mar 2022 12:25:11 GMT

```

It works well for me without any content types. Are you sure you're using the latest Zipkin server?

**Comment by M3yo** [comment]
As I said previously, I'm using Jaeger (docker command on my [first post](https://github.com/spring-cloud/spring-cloud-sleuth/issues/2126#issuecomment-1080727399)) which provides a [Zipkin compatilibity](https://www.jaegertracing.io/docs/1.32/apis/#zipkin-formats-stable)

Indeed Jaeger returns a 400 HTTP error If I don't provide the "Content-Type" header, where as Zipkin server doesn't seem to care about that header.

It feels like the header should be present as we send an POST request with a json body. We can also add the fact that the rest template implementation explicitly set this header, and the webclient one doesn't.

So that clears out the 1st problem with the HTTP Error. 
Still the webclient implementation seems to be buggy, it clearly shows a timeout exception, but my curl command respond in ~5ms. Have you tested my [example](https://github.com/M3yo/sleuth-3.1.1-reactive-bug) app ?

> [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: Timed out after 1000ms}]

**Comment by danieleguia-pro** [comment]
> As I said previously, I'm using Jaeger (docker command on my [first post](https://github.com/spring-cloud/spring-cloud-sleuth/issues/2126#issuecomment-1080727399)) which provides a [Zipkin compatilibity](https://www.jaegertracing.io/docs/1.32/apis/#zipkin-formats-stable)
> 
> Indeed Jaeger returns a 400 HTTP error If I don't provide the "Content-Type" header, where as Zipkin server doesn't seem to care about that header.
> 
> It feels like the header should be present as we send an POST request with a json body. We can also add the fact that the rest template implementation explicitly set this header, and the webclient one doesn't.
> 
> So that clears out the 1st problem with the HTTP Error. Still the webclient implementation seems to be buggy, it clearly shows a timeout exception, but my curl command respond in ~5ms. Have you tested my [example](https://github.com/M3yo/sleuth-3.1.1-reactive-bug) app ?
> 
> > [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: Timed out after 1000ms}]

I rewrite the class WebClientSender and work in jaeger when add header content type
```
package org.springframework.cloud.sleuth.zipkin2;

import java.net.URI;
import org.springframework.http.MediaType;
import org.springframework.web.reactive.function.client.WebClient;
import zipkin2.Span;
import zipkin2.codec.BytesEncoder;
import zipkin2.reporter.Sender;

/**
 * {@link Sender} that uses {@link WebClient} to send spans to Zipkin.
 *
 * @since 3.1.0
 */
public class WebClientSender extends HttpSender {

  public WebClientSender(final WebClient webClient, final String baseUrl, final String apiPath,
      final BytesEncoder<Span> encoder) {

    super((url, mediaType, bytes) -> post(url, mediaType, bytes, webClient), baseUrl, apiPath, encoder);
  }

  private static void post(final String url, final MediaType mediaType, final byte[] json, final WebClient webClient) {

    webClient.post()
        .uri(URI.create(url))
        .header("Content-Type", "application/json")
        .accept(mediaType)
        .bodyValue(json)
        .retrieve()
        .toBodilessEntity()
        .subscribe();
  }

  @Override
  public String toString() {

    return "WebClientSender{" + url + "}";
  }

}
```

**Comment by marcingrzejszczak** [comment]
Are you willing to file a pr with a test?

**Comment by marcingrzejszczak** [comment]
fixed via https://github.com/spring-cloud/spring-cloud-sleuth/commit/22be601b89126bc2a4c212675522a3d021f4b32f

**Comment by patpatpat123** [comment]
Just saw this today with the latest 2021.0.2. Wanted to say thank you @marcingrzejszczak  and thanks everyone on this thread

**Comment by M3yo** [comment]
I've just tested it, unfortunately it's only partially fixed. @marcingrzejszczak (spring-cloud-starter-sleuth 3.1.2)

The missing header "Content-type" is fixed, so I don't get the 400 HTTP error anymore.

But there's still an error with the treatment of the "checkResult" future... I get a TimeoutException whereas if the curl the endpoint, it respond in <10ms.

How to reproduce:

- Start Jaeger/zipkin locally, listening 9411 (default port)
- Start the [application](https://github.com/M3yo/sleuth-3.1.1-reactive-bug) (spring-cloud 2021.0.2)

And the warn is still present.
> 
2022-05-13 14:25:37.598  INFO [,,] 290196 --- [           main] com.example.demo.DemoApplication         : Starting DemoApplication using Java 11.0.2 on ... with PID 290196 (/home/....)
2022-05-13 14:25:37.603  INFO [,,] 290196 --- [           main] com.example.demo.DemoApplication         : No active profile set, falling back to 1 default profile: "default"
2022-05-13 14:25:38.960  INFO [,,] 290196 --- [           main] o.s.cloud.context.scope.GenericScope     : BeanFactory id=d1d2d83b-6fc2-33e4-8438-5bda12280122
2022-05-13 14:25:41.596  WARN [,,] 290196 --- [           main] o.s.c.s.a.z.ZipkinAutoConfiguration      : **Check result of the [WebClientSender{http://localhost:9411/api/v2/spans}] contains an error [CheckResult{ok=false, error=java.util.concurrent.TimeoutException: Timed out after 1000ms}]**
2022-05-13 14:25:42.213  INFO [,,] 290196 --- [           main] o.s.b.web.embedded.netty.NettyWebServer  : Netty started on port 8080
2022-05-13 14:25:42.264  INFO [,,] 290196 --- [           main] com.example.demo.DemoApplication         : Started DemoApplication in 5.593 seconds (JVM running for 6.462)

**Comment by marcingrzejszczak** [comment]
So you're saying that it times out when using WebClient but doesn't time out when using CURL? Is the warn message still present when using Zipkin or it's only Jaeger?

**Comment by M3yo** [comment]
It timesout in both cases (Zipkin or Jaeger), I've just tested it.

If I start with "spring-boot-starter-web" dependency, no issues
If I start with "spring-boot-starter-webflux" dependency instead, I get the timeout

So it's clearly related to the WebClient implementation of the checkResult.

**Comment by LovingCoder** [comment]
Hi, I also encountered this problem, there will be an error using reactor, how did you solve the problem of timeout?
